### PR TITLE
SAAS-3547: Added Kinesis Data Stream Encryption Enabled Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+                    "HashKeyRange": {
+                        "StartingHashKey": "0",
+                        "EndingHashKey": "340282366920938463463374607431768211455"
+                    },
+                    "SequenceNumberRange": {
+                        "StartingSequenceNumber": "49625307919684448350212376424208934206295946525447028738"
+                    }
+                }
+            ],
+            "StreamARN": "arn:aws:kinesis:us-east-1:101363889637:stream/mine2",
```
Comment (sadeed12345) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js:25]:
```suggestion
            "StreamARN": "arn:aws:kinesis:us-east-1:000011112222:stream/mine2",
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            "StreamName": "mine2",
+            "StreamStatus": "ACTIVE",
+            "RetentionPeriodHours": 24,
+            "EnhancedMonitoring": [
+                {
+                    "ShardLevelMetrics": []
+                }
+            ],
+            "EncryptionType": "KMS",
+            "KeyId": "arn:aws:kms:us-east-1:101363889637:key/ad013a33-b01d-4d88-ac97-127399c18b3e",
```
Comment (sadeed12345) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js:35]:
```suggestion
            "KeyId": "arn:aws:kms:us-east-1:000011112222:key/ad013a33-b01d-4d88-ac97-127399c18b3e",
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+                    "HashKeyRange": {
+                        "StartingHashKey": "0",
+                        "EndingHashKey": "340282366920938463463374607431768211455"
+                    },
+                    "SequenceNumberRange": {
+                        "StartingSequenceNumber": "49625308470958869657889380482971889906163549124874469378"
+                    }
+                }
+            ],
+            "StreamARN": "arn:aws:kinesis:us-east-1:101363889637:stream/mine3",
```
Comment (sadeed12345) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js:53]:
```suggestion
            "StreamARN": "arn:aws:kinesis:us-east-1:000011112222:stream/mine3",
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                    "HashKeyRange": {
+                        "StartingHashKey": "0",
+                        "EndingHashKey": "340282366920938463463374607431768211455"
+                    },
+                    "SequenceNumberRange": {
+                        "StartingSequenceNumber": "49625308561945910067894322900437620458568881592741134338"
+                    }
+                }
+            ],
+            "StreamARN": "arn:aws:kinesis:us-east-1:101363889637:stream/mine1",
```
Comment (sadeed12345) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.spec.js:81]:
```suggestion
            "StreamARN": "arn:aws:kinesis:us-east-1:000011112222:stream/mine1",
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Kinesis Data Streams Encrypted',
+    category: 'Kinesis',
+    domain: 'Content Delivery',
+    description: 'Ensures Kinesis Streams are encrypted using keys of desired encryption level',
```
Comment (sadeed12345) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:8]:
```suggestion
    description: 'Ensures Kinesis Streams are encrypted using AWS KMS of desired encryption level',
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Kinesis Data Streams Encrypted',
+    category: 'Kinesis',
+    domain: 'Content Delivery',
+    description: 'Ensures Kinesis Streams are encrypted using AWS KMS of desired encryption level',
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:8]:
```suggestion
    description: 'Ensures Kinesis Streams are encrypted using AWS KMS key of desired encryption level.',
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    more_info: 'Data sent to Kinesis Streams can be encrypted using KMS server-side encryption. Existing streams can be modified to add encryption with minimal overhead.'+
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:9]:
```suggestion
    more_info: 'Data sent to Kinesis Streams can be encrypted using KMS server-side encryption. Existing streams can be modified to add encryption with minimal overhead. '+ 
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Kinesis Data Streams Encrypted',
+    category: 'Kinesis',
+    domain: 'Content Delivery',
+    description: 'Ensures Kinesis Streams are encrypted using AWS KMS of desired encryption level',
+    more_info: 'Data sent to Kinesis Streams can be encrypted using KMS server-side encryption. Existing streams can be modified to add encryption with minimal overhead.'+ 
+        'Use customer-managed keys instead in order to gain more granular control over encryption/decryption process.',
+    recommended_action: 'Enable encryption using desired level for all Kinesis Streams.',
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:11]:
```suggestion
    recommended_action: 'Enable encryption using desired level for all Kinesis streams',
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    link: 'https://docs.aws.amazon.com/streams/latest/dev/server-side-encryption.html',
+    apis: ['Kinesis:listStreams', 'Kinesis:describeStream', 'KMS:listKeys', 'KMS:describeKey'
+        , 'KMS:listAliases', 'STS:getCallerIdentity'],
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:14]:
```suggestion
    apis: ['Kinesis:listStreams', 'Kinesis:describeStream', 'KMS:listKeys', 'KMS:describeKey',
        'KMS:listAliases', 'STS:getCallerIdentity'],
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+                    'Unable to list KMS keys:' + helpers.addError(listKeys), region);
+                return rcb();
+            }   
+
+            var listAliases = helpers.addSource(cache, source,
+                ['kms', 'listAliases', region]);
+
+            if (!listAliases || listAliases.err || !listAliases.data) {
+                helpers.addResult(results, 3,
+                    'Unable to query for KMS aliases:' + helpers.addError(listAliases),
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:72]:
```suggestion
                    'Unable to query for KMS aliases: ' + helpers.addError(listAliases),
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+                    region);
+                return rcb();
+            }
+            
+
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:77]:
```suggestion
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+            var keyArn;
+            var kmsAliasArnMap = {};
+            listAliases.data.forEach(function(alias){
+                keyArn = alias.AliasArn.replace(/:alias\/.*/, ':key/' + alias.TargetKeyId);
+                kmsAliasArnMap[alias.AliasName] = keyArn;
+            });
+
+            for (let stream of listStreams.data) {
+                // arn:aws:kinesis:us-west-1:101363889637:stream/aaaa
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:86]:
```suggestion
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+                
+                var describeStream = helpers.addSource(cache, source,
+                    ['kinesis', 'describeStream', region, stream]);
+
+                if (!describeStream ||
+                    (!describeStream.err && !describeStream.data)) {
+                    continue;
+                }
+
+                if (describeStream.err || !describeStream.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:97]:
```suggestion
                if (!describeStream || describeStream.err || !describeStream.data || !describeStream.data.StreamDescription) {
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+                    helpers.addResult(results, 3,
+                        'Unable to query Kinesis for stream: ' + stream + ': ' + helpers.addError(describeStream),
+                        region);
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:100]:
```suggestion
                        region, resource);
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                    continue;
+                }
+
+                if (!describeStream.data.StreamDescription) {
+                    helpers.addResult(results, 3,
+                        'Unable to query Kinesis for stream: ' + stream + ': no stream data',
+                        region);
+                    continue;
+                }
+
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:110]:
```suggestion
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+                    if (!describeKey || describeKey.err || !describeKey.data || !describeKey.data.KeyMetadata) {
+                        helpers.addResult(results, 3,
+                            `Unable to query KMS key: ${helpers.addError(describeKey)}`,
+                            region, kmsKeyArn);
+                        continue;
+                    }
+                    currentEncryptionLevel = helpers.getEncryptionLevel(describeKey.data.KeyMetadata, helpers.ENCRYPTION_LEVELS);
+
+                } else currentEncryptionLevel = 2; //awskms
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:138]: They don't have default encryption so we should remove this

Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:138]:
```suggestion
```

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+                }
+
+                if (!describeStream.data.StreamDescription.KeyId) {
+                    helpers.addResult(results, 2,
+                        'Kinesis stream does not have encryption enabled',
+                        region, resource);
+                    continue;
+                }
+                
+                if (describeStream.data.StreamDescription.KeyId) {
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:118]:
```suggestion
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/kinesis/kinesisDataStreamsEncrypted.js [hunk 18]
```diff
@@ -0,0 +1,142 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Kinesis Data Streams Encrypted',
+    category: 'Kinesis',
+    domain: 'Content Delivery',
+    description: 'Ensures Kinesis Streams are encrypted using AWS KMS key of desired encryption level.',
```
Comment (AkhtarAmir) [inline plugins/aws/kinesis/kinesisDataStreamsEncrypted.js:8]:
```suggestion
    description: 'Ensures Kinesis data streams are encrypted using AWS KMS key of desired encryption level.',
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**